src/host/outputStream.cpp

```diff
@@ -310,71 +299,75 @@ bool ConhostInternalGetSet::SetRenderMode(const RenderSettings::Mode mode, const
-//     but it is not represented as a function call on out public API surface.
+// - Enables or disables the cursor blinking.
 // Arguments:
 // - fEnable - set to true to enable blinking, false to disable
 // Return Value:
-// - true if successful (see DoSrvPrivateAllowCursorBlinking). false otherwise.
-bool ConhostInternalGetSet::PrivateAllowCursorBlinking(const bool fEnable)
+// - true if successful. false otherwise.
+bool ConhostInternalGetSet::EnableCursorBlinking(const bool fEnable)
 {
-    DoSrvPrivateAllowCursorBlinking(_io.GetActiveOutputBuffer(), fEnable);
+    auto& textBuffer = _io.GetActiveOutputBuffer().GetTextBuffer();
+    textBuffer.GetCursor().SetBlinkingAllowed(fEnable);
+
+    // GH#2642 - From what we've gathered from other terminals, when blinking is
+    // disabled, the cursor should remain On always, and have the visibility
+    // controlled by the IsVisible property. So when you do a printf "\e[?12l"
+    // to disable blinking, the cursor stays stuck On. At this point, only the
+    // cursor visibility property controls whether the user can see it or not.
+    // (Yes, the cursor can be On and NOT Visible)
+    textBuffer.GetCursor().SetIsOn(true);
 
-    bool isPty;
-    DoSrvIsConsolePty(isPty);
     // If we are connected to a pty, return that we could not handle this
     // so that the VT sequence gets flushed to terminal.
-    // Note: we technically don't need to handle it ourselves at all if
-    // we are connected to a pty (i.e. we could have just returned false
-    // immediately without needing to call DoSrvPrivateAllowCursorBlinking),
-    // but we call it anyway for consistency, just in case.
-    return !isPty;
+    return !IsConsolePty();
 }
 
 // Routine Description:
-// - Connects the PrivateSetScrollingRegion call directly into our Driver Message servicing call inside Conhost.exe
-//   PrivateSetScrollingRegion is an internal-only "API" call that the vt commands can execute,
-//     but it is not represented as a function call on out public API surface.
+// - Sets the top and bottom scrolling margins for the current page. This creates
+//     a subsection of the screen that scrolls when input reaches the end of the
+//     region, leaving the rest of the screen untouched.
 // Arguments:
-// - scrollMargins - The bounds of the region to be the scrolling region of the viewport.
+// - scrollMargins - A rect who's Top and Bottom members will be used to set
+//     the new values of the top and bottom margins. If (0,0), then the margins
+//     will be disabled. NOTE: This is a rect in the case that we'll need the
+//     left and right margins in the future.
 // Return Value:
-// - true if successful (see DoSrvPrivateSetScrollingRegion). false otherwise.
-bool ConhostInternalGetSet::PrivateSetScrollingRegion(const SMALL_RECT& scrollMargins)
+// - <none>
+void ConhostInternalGetSet::SetScrollingRegion(const SMALL_RECT& scrollMargins)

```

Comment:
Yeah. All the validation is already handled in `AdaptDispatch::_DoSetTopBottomScrollingMargins`, since that's where the margins are really being managed. This method is just forwarding a copy to `SCREEN_INFORMATION` so they can be accessed from the `AdjustCursorPosition` function. I'm hoping one day we'll be able to avoid that though.